tools/parity_scale_decoder/src/main.rs

```diff
@@ -0,0 +1,279 @@
+//! Parity Scale decoder tool for Iroha data types. For usage run with `--help`
+
+#![allow(clippy::print_stdout, clippy::use_debug, clippy::unnecessary_wraps)]
+
+use std::{collections::BTreeMap, fmt::Debug, fs, io, io::Write, path::PathBuf};
+
+use clap::Parser;
+use eyre::{eyre, Result};
+use iroha_data_model::prelude::*;
+use parity_scale_codec::Decode;
+
+/// Parity Scale decoder tool for Iroha data types
+#[derive(Debug, Parser)]
+#[clap(version, about, author)]
+enum Args {
+    /// Show all available types
+    ListTypes,
+    /// Decode type from binary
+    Decode(DecodeArgs),
+}
+
+#[derive(Debug, clap::Args)]
+struct DecodeArgs {
+    /// Path to the binary with encoded Iroha structure
+    binary: PathBuf,
+    /// Type that is expected to be encoded in binary.
+    /// If not specified then a guess will be attempted
+    #[clap(short, long = "type")]
+    type_id: Option<String>,
+}
+
+/// Function pointer to [`DumpDecoded::dump_decoded()`]
+///
+/// Function pointer is used cause trait object can not be used
+/// due to [`Sized`] bound in [`Decode`] trait
+pub type DumpDecodedPtr = fn(&[u8], &mut dyn Write) -> Result<(), eyre::Error>;
+
+/// Map (Type Name -> `dump_decode()` ptr)
+pub type DumpDecodedMap = BTreeMap<&'static str, DumpDecodedPtr>;
```

Comment:
considering that sdk developers are given a schemas file, I think we should prefer using `IntoSchema::type_name` for keys for types that implement `IntoSchema`

Reply:
That was hard, but I found a good solution insipered by [this](https://doc.rust-lang.org/reference/items/implementations.html#inherent-implementations)

Reply:
That was hard, but I've found a good solution insipered by [this](https://github.com/nvzqz/impls#how-it-works)